model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java

```diff
@@ -1777,55 +1763,34 @@ public void removeComponent(ComponentModel component) {
         if (c == null) return;
         session.users().preRemove(this, component);
         removeComponents(component.getId());
-        em.createNamedQuery("deleteComponentConfigByComponent").setParameter("component", c).executeUpdate();
-        em.remove(c);
+        getEntity().getComponents().remove(c);
     }
 
     @Override
     public void removeComponents(String parentId) {
-        TypedQuery<String> query = em.createNamedQuery("getComponentIdsByParent", String.class)
-                .setParameter("realm", realm)
-                .setParameter("parentId", parentId);
-        List<String> results = query.getResultList();
-        if (results.isEmpty()) return;
-        for (String id : results) {
-            session.users().preRemove(this, getComponent(id));
-        }
-        em.createNamedQuery("deleteComponentConfigByParent").setParameter("parentId", parentId).executeUpdate();
-        em.createNamedQuery("deleteComponentByParent").setParameter("parentId", parentId).executeUpdate();
-
+        getEntity().getComponents().stream()
+                .filter(c -> parentId.equals(c.getParentId()))
+                .forEach(c -> session.users().preRemove(this, getComponent(c.getId())));
```

Comment:
I think you're right. It seems that there is no straightforward way to do it just with a single stream. What do you think about this pattern which IMHO is both readable and reasonably streamed?

```java
@Override
public void removeComponents(String parentId) {
    Predicate<ComponentEntity> sameParent = c -> Objects.equals(parentId, c.getParentId());

    getEntity().getComponents().stream()
            .filter(sameParent)
            .map(this::entityToModel)
            .forEach(c -> session.users().preRemove(this, c));

    getEntity().getComponents().removeIf(sameParent);
}
```